### PR TITLE
feat: secp256k1_blake160_sighash_all support multisig

### PR DESCRIPTION
**Pull request by jjyr**

### CODE REVIEW

**Comment by doitian** [comment]
Please rebase

File: c/dao.c [hunk 1]
```diff
@@ -111,8 +111,13 @@ static int extract_withdraw_header_index(size_t input_index, size_t *index) {
       return ERROR_ENCODING;
     }
   }
-  if (arg_res.attr < 2) {
-    return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+
+  /* last position arg is header index */
+  if (arg_res.attr > 2) {
```
Comment (doitian) [inline c/dao.c]: Why this PR has to update DAO?

File: c/secp256k1_blake160_sighash_all.c [hunk 2]
```diff
@@ -94,70 +115,89 @@ int main(int argc, char* argv[]) {
...
-    blake2b_state blake2b_ctx;
-    blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
-    blake2b_update(&blake2b_ctx, tx_hash, BLAKE2B_BLOCK_SIZE);
-    for (size_t i = 1; i < witness_len; i++) {
+    for (size_t i = 0; i < threshold; i++) {
+      /* Load signature */
       arg_res = mol_cut(&witness_pos, MOL_Witness(i));
       if (arg_res.code != 0) {
-        return ERROR_ENCODING;
+        if (arg_res.attr < threshold) {
```
Comment (doitian) [inline c/secp256k1_blake160_sighash_all.c]: Reusing the field `attr` is really a terrible idea, cc @yangby-cryptape

File: c/secp256k1_blake160_sighash_all.c [hunk 3]
```diff
@@ -94,70 +115,89 @@ int main(int argc, char* argv[]) {
...
+          return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+        } else {
+          return ERROR_ENCODING;
+        }
       }
+      witness_len = arg_res.attr;
```
Comment (doitian) [inline c/secp256k1_blake160_sighash_all.c]: If `witness_len < threshold`, we can return early here.

File: c/secp256k1_blake160_sighash_all.c (hunk 2 above)
Comment (doitian) [reply c/secp256k1_blake160_sighash_all.c]: On error, `attr` is always 0.

File: c/secp256k1_blake160_sighash_all.c [hunk 4]
```diff
@@ -94,70 +115,89 @@ int main(int argc, char* argv[]) {
...
+      blake2b_update(&blake2b_ctx, temp, len);
+      blake2b_final(&blake2b_ctx, temp, BLAKE2B_BLOCK_SIZE);
+
+      uint8_t valid = 0;
+      /* compare with all pubkey hash */
+      for (size_t i = 0; i < sigs_cnt; i++) {
+        if (memcmp(argv[i + 1], temp, BLAKE160_SIZE) != 0) {
+          continue;
+        }
+        if (used_signatures[i] == 1) {
```
Comment (doitian) [inline c/secp256k1_blake160_sighash_all.c]: Check this first before `memcmp`. It is cheaper.

File: c/secp256k1_blake160_sighash_all.c (hunk 3 above)
Comment (jjyr) [reply c/secp256k1_blake160_sighash_all.c]: we already check the length in https://github.com/nervosnetwork/ckb-system-scripts/pull/35/files/dcaffdd0c8e5633e3504ee44352acde53600a5ec#diff-ccd9a3a9b2781124b082417ea9746eb4R125

File: c/secp256k1_blake160_sighash_all.c (hunk 2 above)
Comment (jjyr) [reply c/secp256k1_blake160_sighash_all.c]: So we check `code` before check `attr`.

File: c/dao.c (hunk 1 above)
Comment (jjyr) [reply c/dao.c]: The dao contract assumes the second witness arg is `header index`, which implies we can't use dao with multisig lock, @xxuejie think to use the last arg to indicate `header index` is a good idea, and we can compatible with exists format.

File: c/secp256k1_blake160_sighash_all.c (hunk 2 above)
Comment (doitian) [reply c/secp256k1_blake160_sighash_all.c]: > Reusing the field `attr` is really a terrible idea, cc @yangby-cryptape

We can use the union as a temporary solution.

Comment (yangby-cryptape) [reply c/secp256k1_blake160_sighash_all.c]: > Reusing the field `attr` is really a terrible idea, cc @yangby-cryptape

Split this function into seven functions for different types?

Comment (doitian) [reply c/secp256k1_blake160_sighash_all.c]: change attr to union

```
union {
    uint32_t is_some;
    uint32_t len;
    ...
}
```

File: c/secp256k1_blake160_sighash_all.c (hunk 3 above)
Comment (doitian) [reply c/secp256k1_blake160_sighash_all.c]: In 125, you only check the length when the mol failed to get a slice of witness Vec.

File: c/secp256k1_blake160_sighash_all.c (hunk 2 above)
Comment (doitian) [reply c/secp256k1_blake160_sighash_all.c]: The essential comment of this line is that when `arg_res.code != 0`, `arg_res.attr` is always 0. If `arg_res.code != 0`, we definitely has an encoding error here.

File: c/secp256k1_blake160_sighash_all.c (hunk 3 above)
Comment (doitian) [reply c/secp256k1_blake160_sighash_all.c]: https://github.com/nervosnetwork/ckb-system-scripts/pull/35#discussion_r316133203

Comment (jjyr) [reply c/secp256k1_blake160_sighash_all.c]: I see, the `if` expression is wrong.

File: c/secp256k1_blake160_sighash_all.c (hunk 2 above)
Comment (jjyr) [reply c/secp256k1_blake160_sighash_all.c]: I see, fixed.

Comment (yangby-cryptape) [reply c/secp256k1_blake160_sighash_all.c]: > change attr to union
> 
> ```
> union {
>     uint32_t is_some;
>     uint32_t len;
>     ...
> }
> ```

I got.

**Comment by jjyr** [comment]
@doitian ready for merge

**Pull request merged**